### DISCUSSION TITLE
Receive midi messages from specific device

### DISCUSSION THREAD

**Discussion by elmcapp** [orchetect/swift-midi#99] [Q&A] [answered]
I have one more question regarding receive midi message. How do we get them from a specific device. Currently I am only able to get them from all endpoints. I like to only get message that come from a device named device 1

will give you messages from all endpoints

```swift

try midiManager.addInputConnection(
    toOutputs: [],
    tag: inputTag,
    mode: .allEndpoints,
        receiveHandler: .events { [weak self] events in
            // Note: this handler will be called on a background thread
            // so call the next line on main if it may result in UI updates
            DispatchQueue.main.async {
                events.forEach { self?.receivedMIDIEvent($0) }
            }
        }
    )
```

I tried this to get messages only from device 1 however this did not work for me

```swift
try midiManager.addInputConnection(
    toOutputs: [],
    tag: "device 1",
        receiveHandler: .events { [weak self] events in
            // Note: this handler will be called on a background thread
            // so call the next line on main if it may result in UI updates
            DispatchQueue.main.async {
                events.forEach { self?.receivedMIDIEvent($0) }
            }
        }
    )
```

**Comment by orchetect** [comment, marked as answer]
## If you want all outputs from a specific device:

A device has one or more entities. Entities have one or more endpoints.

- Get `midiManager.devices.devices` which is an array of devices in the system
- Find the device in the array
- Flat-map all `.entities` (property on the device) belonging to the device and their `.outputs` (property on entity)
- Then pass the resulting endpoints into the `.addInputConnection`

## If you want a single specific output:

- Get the list of all output endpoints in the system by getting `midiManager.endpoints.outputs`
- Find the endpoint in the array
- Then pass the endpoint into the `.addInputConnection`

**Reply by elmcapp** [reply]
I'm so sorry but I am completely lost on where to pass the endpoint into the ```.addInputConnection```

only options that I see is filter, mode, toOutput, and tag.

So sorry am lost

**Reply by orchetect** [reply]
Pass the endpoint(s) to `toOutputs:`

**Comment by orchetect** [comment]
Although it's not the ideal way, if you know the name of the endpoint you can do this much simpler. By using the `.name()` criteria, the connection will handle finding the endpoint for you.

I would not do this in production however, since in a system there can be more than one endpoint with the same name. But for testing this can work.

```swift
try midiManager.addInputConnection(
    toOutputs: [.name("device 1")],
    tag: "device 1 connection",
    receiveHandler: .events { [weak self] events in
        // Note: this handler will be called on a background thread
        // so call the next line on main if it may result in UI updates
        DispatchQueue.main.async {
            events.forEach { self?.receivedMIDIEvent($0) }
        }
    }
)
```

**Reply by elmcapp** [reply]
Works perfect. I love your library. It makes handling midi with iOS and macOS very easy. 

Would be nice when you have time to provide an example you would use for production on how you would handle this.

**Reply by orchetect** [reply]
Thanks - I started building the library from scratch around a year and a half ago out of a need for a modern, stable/safe, flexible/modular, and compatible library with complete unit tests that works seamlessly across Apple platforms and supports the newest MIDI spec. It's quite powerful and there's a lot more it can do, but I do this entirely in my free time so I don't have a ton of time to work on it. There are many commercial apps that use the library in production on macOS and iOS so it is well tested.

It may seem complex at first but a lot of time was spent considering API. More examples and documentation eventually will help a lot to ease the learning curve.

Lots of little details with how MIDIKit makes your life easier may not be obvious. Such as when you create a connection, the manager will automatically reconnect to the endpoint(s) if they disappear from the system and reappear again. Like if you unplug a piece of MIDI gear and plug it back in for example.
